packages/opencensus-core/src/common/monitored-resource/gcp-metadata-config.ts

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {get} from 'http';
+import {isString} from 'util';
+
+import * as logger from '../../common/console-logger';
+import * as loggerTypes from '../../common/types';
+
+import {monitoredResourceAttributes} from './types';
+
+export class GcpMetadataConfig {
+  static URL = 'metadata/computeMetadata/v1';
+  static PORT = 80;
+  static readonly HEADER = {'Metadata-Flavor': 'Google'};
+  /** Kubenertes environment variables */
+  static readonly KUBERNETES_SERVICE_HOST = 'KUBERNETES_SERVICE_HOST';
+  static runned = false;
+  /** Persistant metadata */
+  static metadata: Record<string, string> = {};
+  static readonly logger: loggerTypes.Logger = logger.logger();
+  static promise: Promise<Record<string, string>>;
+
+  /**
+   * Initializes metadata service once and load gcp metadata into map.
+   */
+  static run() {
+    if (GcpMetadataConfig.runned) {
+      return GcpMetadataConfig.promise;
+    }
+    GcpMetadataConfig.promise =
+        GcpMetadataConfig.getAttribute('instance_id').then(id => {
+          if (isString(id)) {
+            GcpMetadataConfig.metadata['instance_id'] = id;
+            GcpMetadataConfig.runned = true;
+            return GcpMetadataConfig.getAttributes();
+          }
+          return Promise.resolve(GcpMetadataConfig.metadata);
+        });
+    return GcpMetadataConfig.promise;
+  }
+
+  private static getAttributes() {
+    let attributes;
+    if (GcpMetadataConfig.KUBERNETES_SERVICE_HOST in process.env) {
+      attributes = monitoredResourceAttributes.GKE;
+    } else {
```

Comment:
This is not true. You'll also need to check if `instance-id` exists in the metadata to determine if application is running on GCE.

Reply:
> This is not true. You'll also need to check if `instance-id` exists in the metadata to determine if application is running on GCE.

This method is called just when we have instance-id